languages/julia/reference/progression.md

```diff
@@ -34,6 +34,10 @@ subgraph Type System
 
 	abstract-types --> primitive-types
 end
+
+numbers --> iterator-protocol["iterator-protocol (Fibonacci-Iterator)"]
+multiple-dispatch --> iterator-protocol
+
```

Comment:
These are mainly just notes for myself

Reply:
We should teach the linear time fibonacci algorithm too, maybe, rather than the naive recursive form. I feel like that's more natural for an iterator anyway.

Reply:
> We should teach the linear time fibonacci algorithm too, maybe, rather than the naive recursive form. I feel like that's more natural for an iterator anyway.

My idea was something naïve like

```julia
struct Fib
  to::Int
end

function Base.iterate(fib::Fib, state=(1, 1))
  aₙ = state[1] + state[2]
  aₙ > fib.to && return nothing

  aₙ, (state[2], aₙ)
end
```

```julia
julia> for a in Fib(100)
           println(a)
       end
2
3
5
8
13
21
34
55
89
```

That shows off how to handle the state of an interator.